### ISSUE TITLE
Flutter debugger bug with bool value null

### ISSUE THREAD

**Issue by gregko** [flutter/flutter#17544]
With the following code that triggers exception "Failed assertion: boolean expression must not be null", the exact code line where exception happened is not indicated. When debugging under Android Studio, set breakpoint at the line commented with "// crashes here under debugger!", then press F8 to step through the line. Bam! Crash! Never got to the real offending line just below that. One might think: "What's wrong, does the string contains('...') function have a bug???"

Also remove the above breakpoint and set a breakpoint just below the commented line, where the actual crash actually happens due to testBool being null, then press the FAB + button - the crash happens apparently _before_ this breakpoint is hit, the debugger never stops on the offending line.

Very confusing in a more complex code.

```
class _MyHomePageState extends State<MyHomePage> {
  int _counter = 0;
  static bool testBool;

  @override
  void initState() {
    super.initState();
    SharedPreferences.getInstance().then((sharedPrefs) {
      testBool = sharedPrefs.getBool("testBool"); // testBool will be null
      setState(() {});
    });
  }

  void testBug(String type) {
    String otherStr = '';
    if (type.contains('application/atom+xml')) {
      otherStr = 'abc';
    } else if (type.contains('image/')) { // crashes here under debugger!
      otherStr = testBool ? 'one' : 'two';
      print(otherStr);
    }
  }

  void _incrementCounter() {
    testBug('image/png');
    setState(() {
      _counter++;
    });
  }

  @override
  Widget build(BuildContext context) {...}
}
```
I also have a demo project zipped, don't see anything to attach a file to bug report. You may download the ZIP file instead from:

http://hyperionics.com/downloads/test/bool_null_bug.zip

Flutter doctor output from my machine (I'm working from master channel, but tested dev and beta channels, same problem):

Doctor summary (to see all details, run flutter doctor -v):
[√] Flutter (Channel master, v0.4.4-pre.8, on Microsoft Windows [Version 10.0.16299.431], locale en-US)
[√] Android toolchain - develop for Android devices (Android SDK 27.0.3)
[√] Android Studio (version 3.1)
[!] IntelliJ IDEA Ultimate Edition (version 2018.1)
    X Flutter plugin not installed; this adds Flutter specific functionality.
    X Dart plugin not installed; this adds Dart specific functionality.
[!] VS Code, 32-bit edition (version 1.19.3)
[√] Connected devices (2 available)

! Doctor found issues in 2 categories.

**Comment by jonahwilliams** [comment]
The assertion is generated from the ternary expression in testBug, where testBool is null.  bools in Dart are objects, so they can be null however null itself is not considered 'falsy'.

To avoid situations like this, you can initialize testBool with a default value and avoid setting it if the returned preference is null

**Comment by gregko** [comment]
I learned about this the hard way yesterday, wasting 2 hours due to the bug in the debugger described.

**Comment by jonahwilliams** [comment]
I understand this was frustrating for you - it's very hard to catch errors that one isn't even aware are possible (especially with type names like `bool` instead of `Boolean`).  I would recommend checking out the [dart language tour](https://www.dartlang.org/guides/language/language-tour) for more information.  

Ideally, additions to the type system like nullability and non-null by default could catch errors like this more easily in the future, but we're not quite there yet.

**Comment by gregko** [comment]
Thank you, I did it once, I'll do it again. However I don't think you've read my bug report beyond the first paragraph. So here is the gist: with these two lines, and the problem with testBool value being null actually in the 2nd line below, the debugger crashed in the first one. If I set the  breakpoint on the 2nd line below, the breakpoint was never reached before the crash:

    } else if (type.contains('image/')) { // crashes here under debugger!
      otherStr = testBool ? 'one' : 'two';

This is what was VERY confusing. I thought that there is something wrong with String contains() function, replacing it with other variants (e.g. indexOf(...) > -1) and racking my brain in vain.
